src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesProvisioningLimits.java

```diff
@@ -0,0 +1,133 @@
+
+import javax.annotation.Nonnull;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Implements provisioning limits for clouds and pod templates
+ */
+@Extension
+public final class KubernetesProvisioningLimits {
+    private static final Logger LOGGER = Logger.getLogger(KubernetesProvisioningLimits.class.getName());
+
+    /**
+     * Tracks current number of kubernetes agents per pod template
+     */
+    private final Map<String, AtomicInteger> podTemplateCounts = Collections.synchronizedMap(new HashMap<>());
+
+    /**
+     * Tracks current number of kubernetes agents per kubernetes cloud
+     */
+    private final Map<String, AtomicInteger> cloudCounts = Collections.synchronizedMap(new HashMap<>());
+
+    @Initializer(after = InitMilestone.EXTENSIONS_AUGMENTED)
+    public static void init() {
+        // We don't want anything to be provisioned while we do the initial count.
+        Queue.withLock(() -> {
+            KubernetesProvisioningLimits instance = get();
+            for (Node node : Jenkins.get().getNodes()) {
+                if (node instanceof KubernetesSlave) {
+                    KubernetesSlave kubernetesSlave = (KubernetesSlave) node;
+                    instance.getGlobalCount(kubernetesSlave.getCloudName()).addAndGet(node.getNumExecutors());
+                    instance.getPodTemplateCount(kubernetesSlave.getTemplateId()).addAndGet(node.getNumExecutors());
+                }
+            }
+        });
+    }
+
+    /**
+     * @return the singleton instance
+     */
+    public static KubernetesProvisioningLimits get() {
+        return ExtensionList.lookupSingleton(KubernetesProvisioningLimits.class);
+    }
+
+    /**
+     * Register executors
+     * @param cloud the kubernetes cloud the executors will be on
+     * @param podTemplate the pod template used to schedule the agent
+     * @param numExecutors the number of executors (pretty much always 1)
+     */
+    @SuppressFBWarnings(value="JLM_JSR166_UTILCONCURRENT_MONITORENTER", justification = "Trust me here")
+    public boolean register(@Nonnull KubernetesCloud cloud, @Nonnull PodTemplate podTemplate, int numExecutors) {
+        AtomicInteger globalCount = getGlobalCount(cloud.name);
+        AtomicInteger podTemplateCount = getPodTemplateCount(podTemplate.getId());
+        synchronized (globalCount) {
+            synchronized (podTemplateCount) {
+                if (globalCount.get() + numExecutors <= cloud.getContainerCap()) {
+                    if (podTemplateCount.get() + numExecutors <= podTemplate.getInstanceCap()) {
+                        int g = globalCount.addAndGet(numExecutors);
+                        int p = podTemplateCount.addAndGet(numExecutors);
+                        LOGGER.log(Level.FINEST, () -> cloud.name + ": " + g + "/" + cloud.getContainerCap());
```

Comment:
```suggestion
                        LOGGER.log(Level.FINEST, () -> cloud.name + " global limit: " + g + "/" + cloud.getContainerCap());
```

To be a little more verbose in logs:
```
Feb 01, 2021 4:35:12 PM FINE org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud provision
Template for label "test": test
Feb 01, 2021 4:35:12 PM FINEST org.csanchez.jenkins.plugins.kubernetes.KubernetesProvisioningLimits register
kubernetes: 4/20
Feb 01, 2021 4:35:12 PM FINEST org.csanchez.jenkins.plugins.kubernetes.KubernetesProvisioningLimits register
17540ac3-2f5b-4e1a-9149-4c8933cf8916: 4/10
```